File: Explore/Learn/Introduction to Data Structure/Linked List/Design Linked List/DesignLinkedList.scala
class MyLinkedList() {

    /** Initialize your data structure here. */
    trait MyList[+A]
    case object Nil extends MyList[Nothing]
    case class Node[A](`val`: A, next: MyList[A]) extends MyList[A]
    
    private var head: MyList[Int] = Nil

    /** Get the value of the index-th node in the linked list. If the index is invalid, return -1. */
    def get(index: Int): Int = {
        def aux(node: MyList[Int], index: Int): Int = node match {
            case Nil => -1
            case Node(value, next) => if (index == 0) value else aux(next, index-1)
        }
        aux(head, index)
    }

    /** Add a node of value val before the first element of the linked list. After the insertion, the new node will be the first node of the linked list. */
    def addAtHead(`val`: Int) {
        head = Node(`val`, head)
    }

    /** Append a node of value val to the last element of the linked list. */
    def addAtTail(`val`: Int) {
        def aux(node: MyList[Int], `val`: Int): MyList[Int] = node match {
            case Nil => Node(`val`, Nil)
            case Node(value, next) => Node(value, aux(next, `val`))
        }
        head = aux(head, `val`)
    }

    /** Add a node of value val before the index-th node in the linked list. If index equals to the length of linked list, the node will be appended to the end of linked list. If index is greater than the length, the node will not be inserted. */
    def addAtIndex(index: Int, `val`: Int) {
        def aux(node: MyList[Int], index: Int, `val`: Int): MyList[Int] =
            (node, index) match {
                case (Nil, i) => if (i == 0) Node(`val`, Nil) else Nil
                case (Node(value, next), 0) => Node(`val`, Node(value, next))
                case (Node(value, next), 1) => Node(value, Node(`val`, next))
                case (Node(value, next), i) => Node(value, aux(next, i-1, `val`))
                case (_, _) => node
            }
        head = aux(head, index, `val`)
    }

    /** Delete the index-th node in the linked list, if the index is valid. */
    def deleteAtIndex(index: Int) {
        def aux(node: MyList[Int], index: Int): MyList[Int] =
            (node, index) match {
                case (Nil, _) => Nil
                case (Node(value, next), 0) => next
                case (Node(value1, Node(value2, next)), 1) => Node(value1, next)
                case (Node(value, next), k) => Node(value, aux(next, k-1))
                case (_, _) => node
            }
        head = aux(head, index)
    }

}

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * var obj = new MyLinkedList()
 * var param_1 = obj.get(index)
 * obj.addAtHead(`val`)
 * obj.addAtTail(`val`)
 * obj.addAtIndex(index,`val`)
 * obj.deleteAtIndex(index)
 */

File: Explore/Learn/Introduction to Data Structure/Linked List/Design Linked List/design_linked_list.rs
struct Node {
    val: i32,
    next: Option<Box<Node>>,
}

#[derive(Default)]
struct MyLinkedList {
    head: Option<Box<Node>>,
}


/** 
 * `&self` means the method takes an immutable reference.
 * If you need a mutable reference, change it to `&mut self` instead.
 */
impl MyLinkedList {

    /** Initialize your data structure here. */
    fn new() -> Self {
        Default::default()
    }
    
    /** Get the value of the index-th node in the linked list. If the index is invalid, return -1. */
    fn get(&mut self, index: i32) -> i32 {
        let mut current = match self.head {
            Some(ref a) => a,
            None => return -1,
        };
        
        let mut i = 0;
        while i < index {
            if let Some(ref next) = current.next {
                current = next;
                i += 1;
            } else {
                return -1;
            }
        }
        
        current.val
    }
    
    /** Add a node of value val before the first element of the linked list. After the insertion, the new node will be the first node of the linked list. */
    fn add_at_head(&mut self, val: i32) {
        self.head = Some(Box::new(Node {
            val,
            next: self.head.take(),
        }));
    }
    
    /** Append a node of value val to the last element of the linked list. */
    fn add_at_tail(&mut self, val: i32) {
        let mut current = match self.head {
            Some(ref mut a) => a,
            None => {
                self.head = Some(Box::new(Node {
                    val,
                    next: None,
                }));
                return;
            }
        };
        
        while let Some(ref mut next) = current.next {
            current = next;
        }
        
        current.next = Some(Box::new(Node {
            val,
            next: None,
        }));
    }
    
    /** Add a node of value val before the index-th node in the linked list. If index equals to the length of linked list, the node will be appended to the end of linked list. If index is greater than the length, the node will not be inserted. */
    fn add_at_index(&mut self, index: i32, val: i32) {
        let mut dummy_head = Box::new(Node {
            val: 0,
            next: self.head.take(),
        });
        
        let mut i = 0;
        let mut current = &mut dummy_head;
        while i < index {
            if let Some(ref mut next) = current.next {
                current = next;
            }
            i += 1;
        }
        
        current.next = Some(Box::new(Node {
            val,
            next: current.next.take(),
        }));
        
        self.head = dummy_head.next;
    }
    
    /** Delete the index-th node in the linked list, if the index is valid. */
    fn delete_at_index(&mut self, index: i32) {
        let mut dummy_head = Box::new(Node {
            val: 0,
            next: self.head.take(),
        });
        
        let mut i = 0;
        let mut current = &mut dummy_head;
        while i < index {
            if let Some(ref mut next) = current.next {
                current = next;
            }
            i += 1;
        }
        
        current.next = current.next.take().and_then(|a| a.next);
        
        self.head = dummy_head.next;
    }
}

/**
 * Your MyLinkedList object will be instantiated and called as such:
 * let obj = MyLinkedList::new();
 * let ret_1: i32 = obj.get(index);
 * obj.add_at_head(val);
 * obj.add_at_tail(val);
 * obj.add_at_index(index, val);
 * obj.delete_at_index(index);
 */
